File: c/day2/day2.c
// Advent of Code 2022
// Day 2
// Author: Jacob Christensen

#include <stdio.h>
#include <stdlib.h>

int main(void) {
  
    FILE *f = fopen("input.txt", "r");
    unsigned answer1 = 0;
    unsigned answer2 = 0;
    unsigned char opMove;
    unsigned char myMove;
    unsigned char mvPtr = 0;
    char c;
    int playValues[100];
    playValues['A' - 1] = 3;    // Rollover
    playValues['C' + 1] = 1;
    playValues['A'] = 1;        // Rock
    playValues['X'] = 1;    
    playValues['B'] = 2;        // Paper 
    playValues['Y'] = 2;
    playValues['C'] = 3;        // Scissors
    playValues['Z'] = 3;

    while ((c = getc(f)) != EOF) {

        // Ignore whitespace
        if (c == ' ' || c == '\t') {
            continue;
        }

        if (c == '\n') {
            
            // Part 1
            // Draw
            if (playValues[opMove] == playValues[myMove]) {
                answer1 += playValues[myMove] + 3;
            }
            // Loss
            else if (playValues[opMove - 1] == playValues[myMove]) {
                answer1 += playValues[myMove];
            }
            // Win
            else {
                answer1 += playValues[myMove] + 6;
            }

            // Part 2
            switch (myMove) {
                case 'X': 
                    answer2 += playValues[opMove - 1];
                    break;
                case 'Y':
                    answer2 += playValues[opMove] + 3;
                    break;
                case 'Z':
                    answer2 += playValues[opMove + 1] + 6; 
                    break;
                default:
                    printf("No Match\n");
                    break;
            }

            continue;

        }
        
        if (mvPtr) {
            myMove = c;
            mvPtr--;
        }
        else {
            opMove = c;
            mvPtr++;
        }

    }

    printf("%d\n%d\n", answer1, answer2);

    return EXIT_SUCCESS;

}

File: python/day2/day2.py
# A, X = Rock (1 Point)
# B, Y = Paper (2 Points)
# C, Z = Scissors (3 Points)

WIN_MAP = {'A': 'Y', 'B': 'Z', 'C': 'X'}
DRAW_MAP = {'A': 'X', 'B': 'Y', 'C': 'Z'}
LOSS_MAP = {'A': 'Z', 'B': 'X', 'C': 'Y'}
POINT_MAP = {'X': 1, 'A': 1, 'Y': 2, 'B': 2, 'Z': 3, 'C': 3}
score: int = 0
score1: int = 0

with open('input.txt', 'r') as f:
    games: list = f.readlines()

games = [_.split() for _ in games]

for game in games:
    if WIN_MAP[game[0]] == game[1]:
        score += (POINT_MAP[game[1]] + 6)
    elif DRAW_MAP[game[0]] == game[1]:
        score += (POINT_MAP[game[1]] + 3)
    else:
        score += (POINT_MAP[game[1]])

    match game[1]:
        case 'Z':       # Win
            score1 += (POINT_MAP[WIN_MAP[game[0]]] + 6)
        case 'Y':       # Draw
            score1 += (POINT_MAP[DRAW_MAP[game[0]]] + 3)
        case 'X':       # Lose
            score1 += (POINT_MAP[LOSS_MAP[game[0]]])

print(score, score1)